drivers: dma: intel_lpss: Added parent phandle

akanisetti opened the pull request:
Added a phandle named dma-parent to get base address instead of
adding DMA as child node because it is causing a build warning
(avoid_unnecessary_addr_size) if the parent instance has
"#address-cells/#size-cells" dts properties marked required
and child doesn't have reg property. DMA doesn't have reg
as it gets the base address from parent device.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti committed:
dts: bindings: dma: intel_lpss: Added phandle dma-parent

Added a phandle named dma-parent to get base address instead of
adding DMA as child node because it is causing a build warning
(avoid_unnecessary_addr_size) if the parent instance has
"#address-cells/#size-cells" dts properties marked required
and child doesn't have reg property. DMA doesn't have reg
as it gets the base address from parent device

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti committed:
drivers: dma: intel_lpss: Used phandle to get base address

Added usage of dma_parent phandle instead of using parent-child
method to get DMA base address.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

jhedberg reviewed (changes requested):
@akanisetti thanks for the PR, I see many already approved, but I have a few concerns:

 - It would be good to answer the question "why?" in the commit messages. I don't see any explanation right now for why we can't have DMA nodes as child nodes for peripheral nodes.
 - Does some precedent exist for the `parent-dev` DT property (e.g. in Linux)? Would `dma-parent` have perhaps been better if this is DMA-specific?
 - Normally peripheral nodes have a reference to the DMA node, not vice versa, so you might be creating the possibility for a circular dependency here. Would it not be possible to have a reference to the DMA node from the peripheral node (like is usually the case) and have the peripheral driver initialize the DMA driver (at the same time passing the peripheral device pointer to the DMA driver)?

jhedberg commented:
@akanisetti note that besides the commit message issue, I'm not necessarily requesting other changes. For the other questions I'm also happy with a good explanation (if one exists) :)

akanisetti commented:
> @akanisetti thanks for the PR, I see many already approved, but I have a few concerns:
> 
> * It would be good to answer the question "why?" in the commit messages. I don't see any explanation right now for why we can't have DMA nodes as child nodes for peripheral nodes.
> * Does some precedent exist for the `parent-dev` DT property (e.g. in Linux)? Would `dma-parent` have perhaps been better if this is DMA-specific?
> * Normally peripheral nodes have a reference to the DMA node, not vice versa, so you might be creating the possibility for a circular dependency here. Would it not be possible to have a reference to the DMA node from the peripheral node (like is usually the case) and have the peripheral driver initialize the DMA driver (at the same time passing the peripheral device pointer to the DMA driver)?

1. Updated commit message explaining the reason for the change.
2. Agree with phandle name change as it is more meaningful, updated it.
3. Yes, normally DMA node is provided to peripheral in case of independent DMA but in case of LPSS DMA it depends on the peripheral to get its base address. During init of the DMA itself we need the peripherals reg address and DMA device pointer is needed during the API call of peripheral. Regarding circular dependency, I will be getting the DMAs device node using `device_supported_handles_get` and enabling `DEVICE_DEPS` config.
